plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+        return CheckUnits(**checks)(func)
+    else:
+        return CheckUnits(**checks)
+
+
+def check_values(func=None,
+                 checks_on_return: Dict[str, bool] = None,
+                 **checks: Dict[str, bool]):
+    """
+    A decorator to 'check' -- limit/control -- the values of input and return
+    arguments to a function or method.
+
+    Parameters
+    ----------
+
+    func:
+        The function to be decorated
+
+    checks_on_return: Dict[str, bool]
+        Specifications for value checks on the return of the function being wrapped.
+        (see `check values`_ for valid specifications)
+
+    **checks: Dict[str, Dict[str, bool]]
+        Specifications for value checks on the input arguments of the function
+        being wrapped.  Each keyword argument in `checks` is the name of a function
+        argument to be checked and the keyword value contains the value check
+        specifications.
+
+        .. _`check values`:
+
+        The value check specifications are defined within a dictionary containing
+        the keys defined below.  If the dictionary is empty or omitting keys,
+        then the default value will be assumed for the missing keys.
+
+        ================ ======= ================================================
+        Key              Type    Description
+        ================ ======= ================================================
+        can_be_negative  `bool`  [DEFAULT `True`] values can be negative
+        can_be_complex   `bool`  [DEFAULT `False`] values can be complex numbers
+        can_be_inf       `bool`  [DEFAULT `True`] values can be :data:`~numpy.inf`
+        can_be_nan       `bool`  [DEFAULT `True`] values can be :data:`~numpy.nan`
+        none_shall_pass  `bool`  [DEFAULT `False`] values can be a python `None`
+        ================ ======= ================================================
+
+    Notes
+    -----
+    * Checking of function arguments `*args` and `**kwargs` is not supported.
+    * Full functionality is defined by the class :class:`CheckValues`.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        from plasmapy.utils.decorators import check_values
+
+        @check_values(arg1={'can_be_negative': False, 'can_be_nan': False},
+                      arg2={'can_be_inf': False},
+                      checks_on_return={'none_shall_pass': True)
+        def foo(arg1, arg2):
+            return None
+
+        # on a method
+        class Foo:
+            @check_values(arg1={'can_be_negative': False, 'can_be_nan': False},
+                          arg2={'can_be_inf': False},
+                          checks_on_return={'none_shall_pass': True)
+            def bar(self, arg1, arg2):
+                return None
+    """
+    if checks_on_return is not None:
+        checks['checks_on_return'] = checks_on_return
+
+    if func is not None:
+        return CheckValues(**checks)(func)
+    else:
+        return CheckValues(**checks)
```

Comment:
An alternative (which you can decide if it would be cleaner or not):
```Python
return CheckValues(**checks) if func is None else CheckValues(**checks)(func)
```

Reply:
When the expression gets long enough, I prefer to break the if-statement over multiple lines.  It's just more readable to me.  Of course, this is subjective.

Reply:
Shorter, but a bit more confusing due to the two variants differing only in calling `(func)` at the end. I'd prefer keeping it as it is.